cpp/src/arrow/dataset/dataset_encryption_test.cc

```diff
@@ -0,0 +1,216 @@
+       [ 8, 1, 1, "i" ],
+       [ 9, 0, 2, "j" ]
+     ])"});
+
+    // Use a Hive-style partitioning scheme.
+    partitioning = std::make_shared<HivePartitioning>(schema({field("part", utf8())}));
+
+    // Prepare encryption properties.
+    std::unordered_map<std::string, std::string> key_map;
+    key_map.emplace(kColumnMasterKeysId, kColumnMasterKey);
+    key_map.emplace(kFooterKeyMasterKeyId, kFooterKeyMasterKey);
+
+    crypto_factory = std::make_shared<parquet::encryption::CryptoFactory>();
+    auto kms_client_factory =
+        std::make_shared<parquet::encryption::TestOnlyInMemoryKmsClientFactory>(
+            /*wrap_locally=*/true, key_map);
+    crypto_factory->RegisterKmsClientFactory(std::move(kms_client_factory));
+    kms_connection_config = std::make_shared<parquet::encryption::KmsConnectionConfig>();
+
+    // Set write options with encryption configuration.
+    auto encryption_config =
+        std::make_shared<parquet::encryption::EncryptionConfiguration>(
+            std::string(kFooterKeyName));
+    encryption_config->column_keys = kColumnKeyMapping;
+    auto parquet_encryption_config = std::make_shared<ParquetEncryptionConfig>();
+    parquet_encryption_config->Setup(crypto_factory, kms_connection_config,
+                                     std::move(encryption_config));
+
+    auto file_format = std::make_shared<ParquetFileFormat>();
+    auto parquet_file_write_options =
+        checked_pointer_cast<ParquetFileWriteOptions>(file_format->DefaultWriteOptions());
+    parquet_file_write_options->parquet_encryption_config =
+        std::move(parquet_encryption_config);
+
+    // Write dataset.
+    auto dataset = std::make_shared<InMemoryDataset>(table);
+    EXPECT_OK_AND_ASSIGN(auto scanner_builder, dataset->NewScan());
+    EXPECT_OK_AND_ASSIGN(auto scanner, scanner_builder->Finish());
+
+    FileSystemDatasetWriteOptions write_options;
+    write_options.file_write_options = parquet_file_write_options;
+    write_options.filesystem = file_system;
+    write_options.base_dir = kBaseDir;
+    write_options.partitioning = partitioning;
+    write_options.basename_template = "part{i}.parquet";
+    ASSERT_OK(FileSystemDataset::Write(write_options, std::move(scanner)));
+
+    // Verify that the files exist
+    std::vector<std::string> files = {"part=a/part0.parquet", "part=b/part0.parquet",
+                                      "part=c/part0.parquet", "part=d/part0.parquet",
+                                      "part=e/part0.parquet", "part=f/part0.parquet",
+                                      "part=g/part0.parquet", "part=h/part0.parquet",
+                                      "part=i/part0.parquet", "part=j/part0.parquet"};
+    for (const auto& file_path : files) {
+      ASSERT_OK_AND_ASSIGN(auto result, file_system->GetFileInfo(file_path));
+      ASSERT_EQ(result.type(), fs::FileType::File);
+    }
+  }
+
+ protected:
+  inline static std::shared_ptr<fs::FileSystem> file_system;
```

Comment:
Our convention is to append underscores in members of classes (except for plain light-weight structs). So `file_system_` here, for example.